File: typescript/Behavioral/22-Visitor/Visitor/CostRollupVisitor.ts
import type { IJobVisitor } from "./IJobVisitor";
import type { PlumbingJob } from "../Element/PlumbingJob";
import type { ElectricalJob } from "../Element/ElectricalJob";
import type { CarpentryJob } from "../Element/CarpentryJob";

export class CostRollupVisitor implements IJobVisitor {
  private plumbingTotal = 0;
  private electricalTotal = 0;
  private carpentryTotal = 0;

  get grandTotal(): number {
    return this.plumbingTotal + this.electricalTotal + this.carpentryTotal;
  }

  visitPlumbing(job: PlumbingJob): void {
    this.plumbingTotal += job.cost;
  }

  visitElectrical(job: ElectricalJob): void {
    this.electricalTotal += job.cost;
  }

  visitCarpentry(job: CarpentryJob): void {
    this.carpentryTotal += job.cost;
  }

  printSummary(): void {
    console.log("\n=== Cost Rollup ===");
    console.log(`  Plumbing   : N$${this.plumbingTotal.toFixed(2)}`);
    console.log(`  Electrical : N$${this.electricalTotal.toFixed(2)}`);
    console.log(`  Carpentry  : N$${this.carpentryTotal.toFixed(2)}`);
    console.log("  ─────────────────────────");
    console.log(`  TOTAL      : N$${this.grandTotal.toFixed(2)}`);
  }
}

File: csharp/Behavioral/22-Visitor/Visitor/CostRollupVisitor.cs
namespace Visitor;

public class CostRollupVisitor : IJobVisitor
{
    public decimal PlumbingTotal { get; private set; }
    public decimal ElectricalTotal { get; private set; }
    public decimal CarpentryTotal { get; private set; }

    public decimal GrandTotal =>
        PlumbingTotal + ElectricalTotal + CarpentryTotal;

    public void Visit(PlumbingJob job)   => PlumbingTotal   += job.Cost;
    public void Visit(ElectricalJob job) => ElectricalTotal += job.Cost;
    public void Visit(CarpentryJob job)  => CarpentryTotal  += job.Cost;

    public void PrintSummary()
    {
        Console.WriteLine("\n=== Cost Rollup ===");
        Console.WriteLine($"  Plumbing   : N${PlumbingTotal:N2}");
        Console.WriteLine($"  Electrical : N${ElectricalTotal:N2}");
        Console.WriteLine($"  Carpentry  : N${CarpentryTotal:N2}");
        Console.WriteLine("  ─────────────────────────");
        Console.WriteLine($"  TOTAL      : N${GrandTotal:N2}");
    }
}
